draft-schinazi-masque-h3-datagram.md

```diff
@@ -165,51 +165,71 @@ with error H3_SETTINGS_ERROR.
 
 ~~~
   Datagram-Flow-Id = 2
 ~~~
 
-Definitions of HTTP features that use the "Datagram-Flow-Id" header field MAY
-define their own parameters (parameters are defined in Section 3.1.2 of
-{{STRUCT-FIELD}}). For example, an HTTP method that wishes to use two datagram
-flow identifiers for the lifetime of its request stream could encode the second
-flow identifier as a parameter, which could look like this:
+One element in the list is allowed to be unnamed, but all but one elements
+MUST carry a name. The name of an element is encoded in the key of the first
+parameter of that element (parameters are defined in Section 3.1.2 of
+{{STRUCT-FIELD}}). Each name MUST NOT appear more than once in the list. The
+value of the first parameter of each named element (whose corresponding key
+conveys the element name) MUST be of type Boolean and equal to true. The value
+of the first parameter of the unnamed element MUST NOT be of type Boolean. The
+ordering of the list does not carry any semantics. For example, an HTTP method
+that wishes to use four datagram flow identifiers for the lifetime of its
+request stream could look like this:
 
 ~~~
-  Datagram-Flow-Id = 42; alternate=44
+  Datagram-Flow-Id = 42, 44; ecn-ect0, 46; ecn-ect1, 48; ecn-ce
 ~~~
 
-The "Datagram-Flow-Id" header field MUST NOT be present more than once on a given
-HTTP message; any HTTP message containing more than one "Datagram-Flow-Id"
-header field is malformed.
+In this example, 42 is the unnamed flow identifier, 44 represents the name
+"ecn-ect0", 46 represents "ecn-ect1", and 48 represents "ecn-ce". Note that,
+since the list ordering does not carry semantics, this example can be
+equivalently encoded as:
 
-Since the QUIC STREAM frame that contains the "Datagram-Flow-Id" header field could
-be lost or reordered, it is possible that an endpoint will receive an HTTP/3
-datagram with a flow identifier that it does not know as it has not yet
-received the corresponding "Datagram-Flow-Id" header field. Endpoints MUST NOT treat
-that as an error; they MUST either silently discard the datagram or buffer it
-until they receive the "Datagram-Flow-Id" header field.
+~~~
+  Datagram-Flow-Id = 44; ecn-ect0, 42, 48; ecn-ce, 46; ecn-ect1
+~~~
+
+Even if a sender attempts to communicate the meaning of a flow identifier
+before it uses it in an HTTP/3 datagram, it is possible that its peer will
+receive an HTTP/3 datagram with a flow identifier that it does not know as it
+has not yet received the corresponding "Datagram-Flow-Id" header field. (For
+example, this could happen if the QUIC STREAM frame that contains the
+"Datagram-Flow-Id" header field is reordered and arrives afer the DATAGRAM
+frame.) Endpoints MUST NOT treat that scenario as an error; they MUST either
+silently discard the datagram or buffer it until they receive the
+"Datagram-Flow-Id" header field.
+
+Distinct HTTP requests MAY refer to the same flow identifier in their
+respective "Datagram-Flow-Id" header fields.
```

Comment:
Do flow identifiers have to use the same name (or stay unamed) across requests? I think it would help to be strict here.

Reply:
Discussed offline, the draft needs editorial work to clarify how to use names and their conceptual meaning, but we'll address that in a separate PR.